autogen/browser_utils/markdown_search.py

```diff
@@ -0,0 +1,290 @@
+# ruff: noqa: E722
+import json
+import re
+import requests
+import logging
+import os
+
+from bs4 import BeautifulSoup
+from typing import Any, Dict, List, Optional, Union, Tuple
+from urllib.parse import urlparse, quote, quote_plus, unquote, urlunparse, parse_qs
+from abc import ABC, abstractmethod
+
+from .mdconvert import MarkdownConverter
+
+logger = logging.getLogger(__name__)
+
+
+class AbstractMarkdownSearch(ABC):
+    """
+    An abstract class for providing search capabilities to a Markdown browser.
+    """
+
+    @abstractmethod
+    def __init__(self):
+        pass
+
+    @property
+    @abstractmethod
+    def search(self, query) -> str:
+        pass
+
+
+class BingMarkdownSearch(AbstractMarkdownSearch):
+    """
+    Provides Bing web search capabilities to Markdown browsers.
+    """
+
+    def __init__(self, bing_api_key: str = None, interleave_results: bool = True):
+        """
+        Perform a Bing web search, and return the results formatted in Markdown.
+
+        Args:
+            bing_api_key: key for the Bing search API. If omitted, an attempt is made to read the key from the BING_API_KEY environment variable. If no key is found, BingMarkdownSearch will print a warning, and will fall back to visiting and scraping the live Bing results page. Scraping is objectively worse than using the API, and thus is not recommended.
+            interleave_results: When using the Bing API, results are returned based on category (web, news, videos, etc.), along with instructions for how they should be interleaved on the page. When `interleave` is set to True, these interleaving instructions are followed, and a single results list is returned by BingMarkdownSearch. When `interleave` is set to false, results are separated by category, and no interleaving is done.
+        """
+        super().__init__()
+
+        self._mdconvert = MarkdownConverter()
+        self._interleave_results = interleave_results
+
+        if bing_api_key is None or bing_api_key.strip() == "":
+            self._bing_api_key = os.environ.get("BING_API_KEY")
+        else:
+            self._bing_api_key = bing_api_key
+
+        if self._bing_api_key is None:
+            if not self._interleave_results:
+                raise ValueError(
+                    "No Bing API key was provided. This is incompatible with setting `interleave_results` to False. Please provide a key, or set `interleave_results` to True."
+                )
+
+            logger.warning(
+                "Warning: No Bing API key provided. BingMarkdownSearch will submit an HTTP request to the Bing landing page, but results may be missing or low quality. To resolve this warning, provide a Bing API key by setting the BING_API_KEY environment variable, or using the 'bing_api_key' parameter in by BingMarkdownSearch's constructor. Bing API keys can be obtained via https://www.microsoft.com/en-us/bing/apis/bing-web-search-api\n"
+            )
+
+    def search(self, query: str):
```

Comment:
Can we pass additional arguments to the search? Such as filters?

Reply:
Not at present. We can add that in a future PR. The query can contain the usual bing operators like "site: "